### PR TITLE
DJI RS 2 Sbus Support

### PR DESCRIPTION
**Pull request by JyeSmith**
A huge thanks to Goldeneye Nyzemekka for all the testing 💪 

Please use this PR to comment on how best to add this support in a user friendly way.

How to setup and test.
- Flash your rx with this PR and connect the receivers tx pin to the sbus  pin.
- From the ELRS rx Lua change the protocol to `DJI RS Pro`.
- The ELRS tx should be set to 100Hz Full, and switches to 8ch.
- Channels 1, 2, 4 are pan, tilt, roll
- Aux 1 is unused but should be set as an arm switch and left in the unarmed position.
- Aux 2 should be a 3 way switch. From the switch center position to high/low, Rec start/stop or photos can be captured.
- Aux 3 should be a 3 way switch. Each of the 3 positions corresponds to mode 1, 2, 3.
- Aux 4 should be a 3 way switch. From the switch center position to high/low, Selfie or Recenter can be activated.
- Rec, photos, recenter, and selfies are done by toggling the switch from the center position to high/low, and back to center in a quick 0.4s movement.  With some ETX and mixer magic these functions can be improved and mapped to buttons etc.
- ETX Outputs for the channels should be -100 to 100.

Channel info from @Wodie https://github.com/Wodie/RoninS_CAN/blob/main/Ronin%20S%20S-Bus%20Commands.png

image

### CODE REVIEW

File: src/src/rx-serial/SerialSBUS.cpp [hunk 1]
```diff
@@ -18,22 +21,45 @@ void SerialSBUS::sendLinkStatisticsToFC()
...
+    
+    if (config.GetSerialProtocol() == PROTOCOL_DJI_RS_PRO)
+    {
+        PackedRCdataOut.ch0 = fmap(channelData[0], CRSF_CHANNEL_VALUE_MIN, CRSF_CHANNEL_VALUE_MAX, 352, 1696);
+        PackedRCdataOut.ch1 = fmap(channelData[1], CRSF_CHANNEL_VALUE_MIN, CRSF_CHANNEL_VALUE_MAX, 352, 1696);
+        PackedRCdataOut.ch2 = fmap(channelData[2], CRSF_CHANNEL_VALUE_MIN, CRSF_CHANNEL_VALUE_MAX, 352, 1696);
+        PackedRCdataOut.ch3 = fmap(channelData[3], CRSF_CHANNEL_VALUE_MIN, CRSF_CHANNEL_VALUE_MAX, 352, 1696);
+        PackedRCdataOut.ch4 = fmap(channelData[5], CRSF_CHANNEL_VALUE_MIN, CRSF_CHANNEL_VALUE_MAX, 352, 1696); // Record start/stop and photo
+        PackedRCdataOut.ch5 = fmap(channelData[6], CRSF_CHANNEL_VALUE_MIN, CRSF_CHANNEL_VALUE_MAX, 352, 1696); // Mode
+        PackedRCdataOut.ch6 = fmap(channelData[7], CRSF_CHANNEL_VALUE_MIN, CRSF_CHANNEL_VALUE_MAX, 176,  848); // Recenter and Selfie
```
Comment (JyeSmith) [inline src/src/rx-serial/SerialSBUS.cpp:33]: After some testing the ranges for Mode and Recenter switches appear to be swapped compared to the info from Wodie.  This should be confirmed to still work by more users.

**Comment by JyeSmith** [comment]
How to connect a receiver with the focus wheel.

Sbus output on an ELRS rx is via the TX pad.

image

Move the switch to the sbus position.

image

**Comment by JyeSmith** [comment]
Direct connection requires an addition pull down resistor, and possible regulator to drop the voltage down from 8V to 5V.  **Users should check their receiver max input voltage**.

image

**Comment by JyeSmith** [comment]
Tentatively labeling this for v3.3 because Wez is going to do a heap of testing after this PR is updated with https://github.com/ExpressLRS/ExpressLRS/pull/2186

**Comment by wvarty** [comment]
### Testing progress so far:

**Movement**
Model in ETX is set as:
ch1 = Ail
ch2 = Ele
ch3 = unassigned
ch4 = Rud

Roll, pitch and yaw is controlling the 3 gimbal axes as expected.

**Modes**
Model in ETX is set as:
ch5 = unassigned
ch6 = SA weight 100%
ch7 = SB weight 100%
ch8 = SC weight 100%

- SA and SC channels should be in the middle position by default. This is the "no operation" position for those channels.
- Moving SA up to pos1 for 400ms or less, then back to pos2 triggers the camera to start/stop recording (if connected to the gimbal and in movie mode).
- Moving SA down to pos3 for 400ms or less, then back to pos2 triggers the camera to take a photo (if connected to the gimbal and in photo mode).
- Moving SB to pos1 sets the gimbal in mode-1, pos2 sets the gimbal in mode-2, pos3 sets the gimbal in mode-3
- Moving SC up to pos1 for 400ms or less, then back to pos2 triggers a recenter on the gimbal.
- Moving SC down to pos3 for 400ms or less, then back to pos2 triggers the gimbal to move to "selfie" mode.

**Failsafe**
Failsafe behaviour was:
1. with the failsafe action set to `no pulses` the gimbal stops moving and remains in the last position when handset is powered off.
2. with the failsafe action set to `last position` the gimbal stops moving and remains in the last position when handset is powered off (i.e. same as `no pulses`). The gimbal appears to be reading the SBUS failsafe bit, which means the failsafe action is the same for both settings.

File: src/src/rx-serial/SerialSBUS.cpp (hunk 1 above)
Comment (wvarty) [reply src/src/rx-serial/SerialSBUS.cpp:45]: Seems to work as-is here in the PR, with the swapped endpoints.

Tested with the endpoints from Woodie, and the recenter/selfie channel still works, but when you trigger the selfie mode, the gimbal mode also toggles from whatever its currently set to, and changes to mode 3 momentarily.

So for example, you have the mode channel set in the middle pos, and the gimbal is in mode 2, then you trigger selfie, the mode changes to mode 3 for a sec, then back to mode 2.

With the channel endpoints as they are in the PR, this does not happen, and the mode stays as set.

**Review by wvarty — APPROVED** [review]
Re-tested after the rebase with `master` and everything is looking good.

**Review by pkendall64 — APPROVED** [review]
Harrumpf!

**Comment by DillanS212** [comment]
Some gimbal users are currently using an SBUS servo to lock and unlock these gimbals by physically pushing the lock/unlock button on the gimbal. 

Can the AUX 1 channel be used for this, or does this receiver protocol output anything on that channel?

**Comment by JyeSmith** [comment]
@DillanS212 all other channels are still usable.  But because aux1 is a special case with ELRS I moved it to sbus ch16.  All of the other channel were also moved down one e.g. aux2 on sbus ch5, aux3 to ch6, and so on.

**Comment by Slikster** [comment]
Is there any way to lock the gymbal through the sbus link? 
using a servo to hit the power button is really crude :P

**Comment by JyeSmith** [comment]
@Slikster not than I'm aware of right now. It would be great to get some official documentation from DJI and know for sure.

**Comment by hsmw123** [comment]
我用tx16s只能控制云台的运动实现不了回中和录制。是什么问题

**Comment by przfpv** [comment]
> @Slikster not than I'm aware of right now. It would be great to get some official documentation from DJI and know for sure.

Has this been figured out yet? I have seen crossfire modules control “sleep mode/motor pause” aka gimbal lock via sbus. Wondering if the same could be done via ELRS

**Comment by mexxxxxx266** [comment]
Hey guess, is there a way to Zoom? With a Sony and a pz lense? Is there a Channel for this?

**Pull request merged**